### PR TITLE
Hash keys for data tries

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- In order to auto balance data tries, when inserting in one insert at hash(key) instead of the key
  
## Proposed changes
- When inserting in a data trie, insert  at `hash(key)`. When getting from a data trie, firts try to retrieve the val from `hash(key)`, and if that does not work, try and get from `key`. 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/delegation_test.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/testscommon/enableEpochsHandlerMock"
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation_test.go:8]: go imports

File: state/trackableDataTrie.go [hunk 2]
```diff
@@ -114,26 +156,102 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) (map[string][
...
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tdaw.dirtyData = make(map[string][]byte)
+	return oldValues, nil
+}
+
+func (tdaw *trackableDataTrie) getOldKeyAndVal(key string) ([]byte, []byte, error) {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply state/trackableDataTrie.go:215]: or getOldKeyAndValWithCleanup ?

File: state/trackableDataTrie.go [hunk 3]
```diff
@@ -114,26 +156,102 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) (map[string][
...
 	}
 
+	if tdaw.enableEpochsHandler.IsAutoBalanceDataTriesEnabled() {
+		return tdaw.updateTrieWithAutoBalancing()
+	}
+
+	return tdaw.updateTrieNormal()
+}
+
+func (tdaw *trackableDataTrie) updateTrieNormal() (map[string][]byte, error) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:166]: Done.

File: state/trackableDataTrie.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:215]: Renamed using @AdoAdoAdo's suggestion.

File: state/trackableDataTrie.go [hunk 4]
```diff
@@ -114,26 +156,102 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) (map[string][
...
+
+	err = tdaw.tr.Delete([]byte(key))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return []byte(key), oldVal, nil
+}
+
+func (tdaw *trackableDataTrie) updateTrie(key []byte, val []byte) error {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:240]: The trie is modified inside this func (`tdaw.tr.Update()` is called), I think `updateTrie` is a better definition of what the func does.

File: vm/systemSmartContracts/delegation_test.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply vm/systemSmartContracts/delegation_test.go:8]: Done.

File: trie/patriciaMerkleTrie.go [hunk 5]
```diff
@@ -585,6 +586,18 @@ func (tr *patriciaMerkleTrie) VerifyProof(rootHash []byte, key []byte, proof [][
 	tr.mutOperation.Lock()
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:586]: Yes, checked this func and indeed, there is nothing that changes inside the trie.

File: state/trackableDataTrie.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/trackableDataTrie.go:223]: yes but we do not actually update the trie (pointer) of the trackable data trie. I still feel this is somehow misleading

File: state/userAccount.go [hunk 6]
```diff
@@ -169,6 +176,37 @@ func (a *userAccount) SetCodeMetadata(codeMetadata []byte) {
 	a.CodeMetadata = codeMetadata
 }
 
+// GetAllLeaves returns all the leaves of the account's data trie
+func (a *userAccount) GetAllLeaves(
+	leavesChannels *common.TrieIteratorChannels,
+	ctx context.Context,
+) error {
+	if check.IfNil(a.marshaller) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/userAccount.go:184]: Please refactor this. We should not work with nil components & subcomponents. 
The solution here is to actually have a different implementation for the empty user account that will override the GetAllLeaves function and return an ErrNotImplemented method.

[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/userAccount.go:184]: Refactored.

**Review by iulianpascalau — APPROVED** [review]

File: common/trie.go [hunk 7]
```diff
@@ -15,3 +19,17 @@ func IsEmptyTrie(root []byte) bool {
...
 }
+
+// TrimSuffixFromValue returns the value without the suffix
+func TrimSuffixFromValue(value []byte, suffixLength int) ([]byte, error) {
+	if suffixLength == 0 {
+		return value, nil
+	}
+
+	dataLength := len(value) - suffixLength
+	if dataLength < 0 {
```
Comment (AdoAdoAdo) [inline common/trie.go:30]: can datalen be 0?

**Review by AdoAdoAdo — DISMISSED** [review]

File: common/trie.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply common/trie.go:30]: No

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**